blur/rand.py

```diff
@@ -413,42 +413,43 @@ def weighted_shuffle(weights):
     Raises:
         ValueError: If passed ``weights`` is not formed correctly
     """
-    working_list = weights[:]
-    # list of tuples [list_item, target_index]
-    # used to store positions before moving them into the final list
-    shuffle_positions = []
-
-    def closest_available(requested_index):
-        """The closest index that isn't taken in ``shuffle_positions``."""
-        taken_positions = [position[1] for position in shuffle_positions]
-        available_indexes = [i for i in range(len(weights))
-                             if i not in taken_positions]
-        return min((available_index for available_index in available_indexes),
-                   key=lambda index: abs(move_index - index))
-
-    while working_list:
-        # Pick which item to move
-        move_index = weighted_choice(
-            [(index, weight[2]) for index, weight in enumerate(working_list)])
-        move_item = working_list[move_index][0]
-        # Find the index where the item will be placed
-        if isinstance(working_list[move_index][1], str):
-            if working_list[move_index][1] == 'STAY':
-                # Place in the index closest to where the item appears already
-                target_position = closest_available(move_index)
+
+    def closest_available(requested_index, available_indices):
+        """The closest index to requested_index that is in the list of available_indices"""
+        # shuffles the available indexes to remove bias when finding the closest one
+        random.shuffle(available_indices)

```

Comment:
Would the shuffle in 422 take care of any bias introduced here from same-weight items?